core/trino-main/src/main/java/io/trino/cost/TopNRankingStatsRule.java

```diff
@@ -0,0 +1,246 @@
+                .setOutputRowCount(outputRowsCount)
+                .addSymbolStatistics(node.getRankingSymbol(), SymbolStatsEstimate.builder()
+                        .setLowValue(1)
+                        // maxRankingPerPartition is the maximum rank per partition. Therefore, in the worse case
+                        // scenario the rank symbol can have maxRankingPerPartition as the high value.
+                        .setHighValue(node.getMaxRankingPerPartition())
+                        .setDistinctValuesCount(rankDistinctCount)
+                        .setNullsFraction(0.0)
+                        .setAverageRowSize(BIGINT.getFixedSize())
+                        .build())
+                .build());
+    }
+
+    private static void adjustOrderBySymbolDistinctCount(
+            List<Symbol> orderBy,
+            double partitionCount,
+            double topRowsPerPartition,
+            double averageRowsPerPartition,
+            PlanNodeStatsEstimate sourceStats,
+            PlanNodeStatsEstimate.Builder estimateBuilder)
+    {
+        verify(!orderBy.isEmpty(), "Order by symbols should not be empty for TopNRankingNode.");
+        Symbol firstSortSymbol = orderBy.getFirst();
+        SymbolStatsEstimate symbolStats = sourceStats.getSymbolStatistics(firstSortSymbol);
+        // We assume that the order by columns are uncorrelated to partition columns, hence order by distinct rows
+        // should be equally spread among partitions.
+        double symbolDistinctCount = min(symbolStats.getDistinctValuesCount(), averageRowsPerPartition);
+        // Consider that order by columns are COL1, COL2 for the below single partition:
+        // COL1   COL2
+        // A      1
+        // A      2
+        // A      3
+        // A      4
+        // B      5
+        // B      6
+        // B      7
+        // B      8
+        // The max rows per partition is 8, and for rank 4, the top rows per partition is 4.
+        // So, in that case, the new distinct count for COL1 should be 1. We can calculate the new distinct count
+        // per partition using the formula:
+        // (symbolStats.getDistinctValuesCount() / averageRowsPerPartition) * topRowsPerPartition => (2 / 8) * 4 = 1
+        // Additionally, it's quite hard to argue about for COL2, thus we will not change the distinct count for COL2.
+        // It'll automatically be handled by the StatsNormalizer based on outputRowCount.
+        double newDistinctCountPerPartition = ceil((symbolDistinctCount / averageRowsPerPartition) * topRowsPerPartition);
+        if (isNaN(newDistinctCountPerPartition)) {
+            return;
+        }
+
+        // Since there's an assumption that order by columns are uncorrelated to partition columns which means
+        // order by distinct rows should be equally spread among partitions, therefore we can simply set
+        // newDistinctCount across all partitions to newDistinctCountPerPartition.
```

Comment:
nit: `newDistinctCount across all partitions to newDistinctCountPerPartition, because each partition will contain same distinct rows according to ordering`